commit a18f92501d3d74bdc8b7428e4201056b0381d422 (refs/changes/48/848/3)
Author: Chris Burroughs <chris.burroughs@joyent.com>
Date:   2016-11-04T17:19:23-04:00 (2 years, 11 months ago)
    
    joyent/node-manta#272 proudly proclaim the latest version
    joyent/node-manta#282 rely on dashdash for mchmod arg parsing

diff --git a/bin/mchattr b/bin/mchattr
index 57a8e60..f759774 100755
--- a/bin/mchattr
+++ b/bin/mchattr
@@ -60,6 +60,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mchmod b/bin/mchmod
index aadd417..02dd805 100755
--- a/bin/mchmod
+++ b/bin/mchmod
@@ -40,23 +40,15 @@ function ifError(err) {
 }
 
 
-function parseMchmodOptions() {
-    if (process.argv.length <= 2) {
+function parseMchmodOptions(args) {
+    if (args.length < 2) {
         manta.cli_usage(OPTIONS_PARSER, 'role required',
                                         '-- [+-=]role,... path...');
     }
 
-    var startRoles = process.argv.indexOf('--');
-    var operation;
-    var roleOpt;
+    var roleOpt = args.splice(0, 1).pop();
+    var operation = roleOpt.charAt(0);
 
-    if (startRoles < 0) {   // no other options: `mchmod +role path`
-        roleOpt = process.argv.splice(2, 1).pop();
-    } else {                // other options: `mchmod -a account -- +role path`
-        roleOpt = process.argv.splice(startRoles, 2).pop();
-    }
-
-    operation = roleOpt.charAt(0);
     if (operation !== '+' && operation !== '-' && operation !== '=') {
         manta.cli_usage(OPTIONS_PARSER,
             'operation should be one of "+", "-" or "="');
@@ -72,8 +64,6 @@ function parseMchmodOptions() {
 function parseOptions() {
     var opts;
 
-    var mchmodOpts = parseMchmodOptions();
-
     try {
         opts = OPTIONS_PARSER.parse(process.argv);
         manta.checkBinEnv(opts);
@@ -82,9 +72,6 @@ function parseOptions() {
                                         '-- [+-=]role,... path...');
     }
 
-    opts.mchmodOperation = mchmodOpts.operation;
-    opts.mchmodRoles = mchmodOpts.roles;
-
     manta.cli_logger(opts, LOG);
 
     if (opts.help) {
@@ -92,6 +79,12 @@ function parseOptions() {
                                         '-- [+-=]role,... path...');
     }
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
+    var mchmodOpts = parseMchmodOptions(opts._args);
+    opts.mchmodOperation = mchmodOpts.operation;
+    opts.mchmodRoles = mchmodOpts.roles;
+
     if (opts._args.length < 1) {
         manta.cli_usage(OPTIONS_PARSER, 'path required',
                                         '-- [+-=]role,... path...');
diff --git a/bin/mfind b/bin/mfind
index c213f2a..ecc71f8 100755
--- a/bin/mfind
+++ b/bin/mfind
@@ -109,6 +109,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, '[PATH]');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts.name)
         opts.name = new RegExp(opts.name);
 
diff --git a/bin/mget b/bin/mget
index 2015526..ed6ec1a 100755
--- a/bin/mget
+++ b/bin/mget
@@ -82,6 +82,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/minfo b/bin/minfo
index 3876c37..11dfa67 100755
--- a/bin/minfo
+++ b/bin/minfo
@@ -53,6 +53,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mjob b/bin/mjob
index e252c56..4d25f11 100755
--- a/bin/mjob
+++ b/bin/mjob
@@ -46,6 +46,8 @@ function assertOpts(thisp, subcmd, opts, cb) {
         return (false);
     }
 
+    manta.handleOptsVersion(opts);
+
     try {
         manta.checkBinEnv(opts);
         manta.cli_logger(opts, LOG);
@@ -429,11 +431,26 @@ function getJobInfo(opts, cb) {
 function MJob() {
     cmdln.Cmdln.call(this, {
         name: 'mjob',
-        desc: 'Manages Manta compute jobs'
-    });
+        desc: 'Manages Manta compute jobs',
+        options: [
+            {
+                names: ['help', 'h'],
+                type: 'bool',
+                help: 'Show this help message and exit.'
+            },
+            {
+                names: ['version'],
+                type: 'bool',
+                help: 'Print version and exit.'
+            }
+        ]});
 }
 util.inherits(MJob, cmdln.Cmdln);
 
+MJob.prototype.init = function (opts, args, callback) {
+    manta.cliVersionCheckPrintAndExit(opts);
+    cmdln.Cmdln.prototype.init.apply(this, arguments);
+};
 
 //-- AddKeys
 
diff --git a/bin/mln b/bin/mln
index 1dc3164..0122d0f 100755
--- a/bin/mln
+++ b/bin/mln
@@ -61,6 +61,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'source dest');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'source required', 'source dest');
     if (opts._args.length < 2)
diff --git a/bin/mlogin b/bin/mlogin
index e131a77..499709e 100755
--- a/bin/mlogin
+++ b/bin/mlogin
@@ -124,6 +124,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length > 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mls b/bin/mls
index f9ed516..f8d0c5c 100755
--- a/bin/mls
+++ b/bin/mls
@@ -94,6 +94,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, '[PATH]...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     opts.paths = [];
     if (opts._args.length < 1) {
         opts.paths = ['/' + opts.account + '/stor'];
diff --git a/bin/mmd5 b/bin/mmd5
index 165559a..75b0333 100755
--- a/bin/mmd5
+++ b/bin/mmd5
@@ -56,6 +56,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mmkdir b/bin/mmkdir
index 10dd55a..ece7f56 100755
--- a/bin/mmkdir
+++ b/bin/mmkdir
@@ -71,6 +71,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     opts.headers = {};
     (opts.header || []).forEach(function (h) {
         if (h.indexOf(':') === -1) {
diff --git a/bin/mput b/bin/mput
index 991344b..e76f45a 100755
--- a/bin/mput
+++ b/bin/mput
@@ -113,6 +113,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts.md5 && !opts.file)
         manta.cli_usage(OPTIONS_PARSER, false, '--md5 requires --file');
 
diff --git a/bin/mrm b/bin/mrm
index 9925fa1..1ba4b67 100755
--- a/bin/mrm
+++ b/bin/mrm
@@ -66,6 +66,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/mrmdir b/bin/mrmdir
index f3e7768..f145d5f 100755
--- a/bin/mrmdir
+++ b/bin/mrmdir
@@ -51,6 +51,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (opts._args.length < 1)
         manta.cli_usage(OPTIONS_PARSER, 'path required', 'path...');
 
diff --git a/bin/msign b/bin/msign
index 6e3d490..789c2d1 100755
--- a/bin/msign
+++ b/bin/msign
@@ -89,6 +89,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path...');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     opts.method = opts.method.map(function (m) {
         m = m.toUpperCase();
         if (METHODS.indexOf(m) === -1) {
diff --git a/bin/muntar b/bin/muntar
index 41d6060..b830b2d 100755
--- a/bin/muntar
+++ b/bin/muntar
@@ -100,6 +100,8 @@ function parseOptions() {
     if (opts.help)
         manta.cli_usage(OPTIONS_PARSER, false, 'path');
 
+    manta.cliVersionCheckPrintAndExit(opts);
+
     if (!opts.file)
         manta.cli_usage(OPTIONS_PARSER, 'file is a required argument', 'path');
 
diff --git a/lib/create_client.js b/lib/create_client.js
index 86b753b..9d13045 100644
--- a/lib/create_client.js
+++ b/lib/create_client.js
@@ -11,6 +11,7 @@ var restifyClients = require('restify-clients');
 var auth = require('smartdc-auth');
 var manta = require('./client');
 
+var packageJson = require('../package.json');
 
 
 ///--- Globals
@@ -42,6 +43,11 @@ var DEFAULT_OPTIONS = [
         type: 'bool',
         help: 'Print this help and exit'
     },
+    {
+        name: 'version',
+        type: 'bool',
+        help: 'Print version and exit.'
+    },
     {
         names: ['insecure', 'i'],
         type: 'bool',
@@ -272,6 +278,13 @@ function setupLogger(opts, log) {
 }
 
 
+function versionCheckPrintAndExit(opts) {
+    if (opts.version) {
+        console.log(packageJson.version);
+        console.log(packageJson.description + ': <' + packageJson.homepage + '>');
+        process.exit(0);
+    }
+}
 
 ///--- Exports
 
@@ -283,7 +296,8 @@ module.exports = {
     createBinClient: createBinClient,
     DEFAULT_OPTIONS: DEFAULT_OPTIONS,
     usage: usage,
-    setupLogger: setupLogger
+    setupLogger: setupLogger,
+    versionCheckPrintAndExit: versionCheckPrintAndExit
 };
 
 
diff --git a/lib/index.js b/lib/index.js
index da64fb1..8fb1acd 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -43,6 +43,7 @@ module.exports = {
     DEFAULT_CLI_OPTIONS: cc.DEFAULT_OPTIONS,
     cli_usage: cc.usage,
     cli_logger: cc.setupLogger,
+    cliVersionCheckPrintAndExit: cc.versionCheckPrintAndExit,
     StringStream: StringStream,
     path: manta.path,
     jobPath: manta.jobPath,
